lib/AST/ASTImporter.cpp

```diff
@@ -2591,14 +2616,30 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
   if (ImportTemplateInformation(D, ToFunction))
     return nullptr;
 
+  bool IsFriend = D->isInIdentifierNamespace(Decl::IDNS_OrdinaryFriend);
+
   // Add this function to the lexical context.
   // NOTE: If the function is templated declaration, it should be not added into
   // LexicalDC. But described template is imported during import of
   // FunctionTemplateDecl (it happens later). So, we use source declaration
   // to determine if we should add the result function.
-  if (!D->getDescribedFunctionTemplate())
+  if (!D->getDescribedFunctionTemplate() && !IsFriend)
     LexicalDC->addDeclInternal(ToFunction);
 
+  // Friend declarations lexical context is the befriending class, but the
+  // semantic context is the enclosing scope of the befriending class.
+  // We want the friend functions to be found in the semantic context by lookup.
+  // FIXME should we handle this genrically in VisitFriendDecl?
+  // In Other cases when LexicalDC != DC we don't want it to be added,
+  // e.g out-of-class definitions like void B::f() {} .
+  if (LexicalDC != DC && IsFriend) {
+    DC->makeDeclVisibleInContext(ToFunction);
+  }
+
+  // Import the rest of the chain. I.e. import all subsequent declarations.
+  for (++RedeclIt; RedeclIt != Redecls.end(); ++RedeclIt)
```

Comment:
Should we do this before or after importing overrides or it does not matter?

Reply:
Does not matter, it should be orthogonal, but it makes sense to handle the overrides once the whole chain is imported.